### PR TITLE
MR / YARN application is not killed when HadoopJavaJob is killed if azkaban.flow.execid is overridden (#1815)

### LINKED ISSUE #1815: MR / YARN application is not killed when HadoopJavaJob is killed if azkaban.flow.execid is overridden
Azkaban writes the job logs to a file `_job.<execId>.<jobId>.log`. If a job is killed, Azkaban scans this log file for YARN applications it should kill. However, if `azkaban.flow.execid` is overridden (either set as a job property or overridden as a flow parameter), then if the Azkaban job is killed, when `HadoopJavaJob.cancel()` is called, it encounters the error

```
java.lang.IllegalArgumentException: the logFilePath does not exist
```

and the launched YARN applications are not killed.

To fix this, when constructing the log file name in the `cancel()` method, it should use the original Azkaban-generated execution id, which can be found in the JVM parameters (`JobRunner.insertJVMAargs()` adds the original execution id as a system property).

Internal JIRA ref: LIHADOOP-37904

### PR DESCRIPTION
**Pull request by erwa**
Fixes #1815.

With this change, even if a user overrides `azkaban.flow.execid`, the original execution id Azkaban generated will still be used to find the log file, which is searched for YARN applications to kill when `HadoopJavaJob.cancel()` is called.

### CODE REVIEW

**Comment by oliverhu** [comment]
What's the point of providing the capability to override _azkaban.flow.execid_ ?

File: az-hadoop-jobtype-plugin/src/main/java/azkaban/jobtype/HadoopJavaJob.java [hunk 1]
```diff
@@ -249,6 +249,22 @@ public String toString() {
         + getJobProps() + '}';
   }
 
+  /**
+   * Returns the {@code azkaban.execid} system property, if set, and otherwise returns the value of
+   * {@value CommonJobProperties#EXEC_ID}. {@value CommonJobProperties#EXEC_ID} may be overridden by a flow param or
+   * job property, whereas the {@code azkaban.execid} system property will be the unique execution id Azkaban generates
+   * for each flow.
+   */
+  private String getExecId() {
```
Comment (oliverhu) [inline az-hadoop-jobtype-plugin/src/main/java/azkaban/jobtype/HadoopJavaJob.java]: I'm not sure if we want to piggyback on this tho, what if user intentionally wants to kill the exec_id he overrides with?

**Comment by erwa** [comment]
Hey Oliver,

The use case for overriding the execution id was as follows:

* A flow writes data to HDFS at a location configured via a parameter `output_location=/tmp/${azkaban.flow.execid}`.
* In one run (execution id `123`), the flow fails on job 3.
* In the next run (execution id `124`), instead of re-running all the jobs, the user just wants to run job 3 and read the data already generated, but it is located at `/tmp/123` rather than `/tmp/124`. So he overrides `${azkaban.flow.execid}` and sets it to `123`.

A simple workaround that does not require overriding Azkaban runtime properties is to define another property:

```
guid=${azkaban.flow.execid}
output_location=/tmp/${guid}
```

Then the user can just override `guid` and leave `azkaban.flow.execid` untouched.

I've also filed #1818 for discussing whether Azkaban should allow Azkaban-injected runtime properties to be overwritten or not.

File: az-hadoop-jobtype-plugin/src/main/java/azkaban/jobtype/HadoopJavaJob.java (hunk 1 above)
Comment (erwa) [reply az-hadoop-jobtype-plugin/src/main/java/azkaban/jobtype/HadoopJavaJob.java]: User does not control the `cancel()` method that kills the job. The `cancel()` method should be able to find the log file that Azkaban created -- the log file is then searched for YARN application that need to be killed.

**Comment by chengren311** [comment]
@erwa thanks for the PR.
 In this case, user job should have a parameter for input data location. Overriding flow execution id itself seems a bit odd to me. azkaban runtime properties by concept are generated by azkaban platform and immutable to users.

**Comment by erwa** [comment]
Thanks for the discussion, Cheng. Upon further reflection, I think using the workaround is a cleaner solution. Will discard this pull request.

**Pull request closed without merging**